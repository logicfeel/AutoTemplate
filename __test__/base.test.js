const fs        = require("fs");
const path      = require("path");
const AutoTask  = require("../src/auto-task").AutoTask;
const dirname   = path.join(__dirname, "/base/mod1");
let autoTask    = null;

/**
 * 가짜 경로로 변경
 * @param {*} obj 
 * @returns 
 */
function changeFakePath(obj) {
    const dir = obj.dir;
    const fake = "@@";

    if (obj.origin) {
        return {
            isPublic: obj.isPublic,
            content: obj.content,
            dir: obj.dir.replace(dir, fake),
            area: obj.area,
            fullPath: obj.fullPath.replace(dir, fake),
            areaDir: obj.areaDir,
            subDir: obj.subDir,
            subPath: obj.subPath,
            localDir: obj.localDir,
            localPath: obj.localPath,
            name: obj.name,
            fileName: obj.fileName,
            filePath: obj.filePath.replace(dir, fake),
            saveName: obj.saveName,
            saveDir: obj.saveDir.replace(dir, fake),
            savePath: obj.savePath.replace(dir, fake),
            origin: obj.origin,
        };
    } else {
        return {
            isPublic: obj.isPublic,
            content: obj.content,
            dir: obj.dir.replace(dir, fake),
            area: obj.area,
            fullPath: obj.fullPath.replace(dir, fake),
            areaDir: obj.areaDir,
            subDir: obj.subDir,
            subPath: obj.subPath,
            localDir: obj.localDir,
            localPath: obj.localPath,
            name: obj.name,
            fileName: obj.fileName,
            filePath: obj.filePath.replace(dir, fake),
        };
    }
}

describe("task :: clear", () => {
    it("[ new >> do_clear(1) ]", () => {
        autoTask = AutoTask.create(dirname);
        autoTask.isLog = false;
        autoTask.do_clear(1);   // 강제 클리어
    });
    it("- 파일 유무 : src/one.html(X)", () => {
        const fullPath = dirname + "/src/one.html";
        expect(fs.existsSync(fullPath)).toBeFalsy();
        
    });
});

describe("task :: publish", () => {
    it("[ new >> do_publish() ]", () => {
        autoTask = AutoTask.create(dirname);
        autoTask.isLog = false;
        autoTask.do_publish();
    });
    it("- 파일 유무 : src/one.html", () => {
        const fullPath = path.join(dirname, "/src/one.html");
        expect(fs.existsSync(fullPath)).toBeTruthy();
    });
    it("- 파일 비교 : src/one.html", () => {
        const fullPath = path.join(dirname, "/src/one.html");
        const data = fs.readFileSync(fullPath,"utf-8");
        expect(fs.existsSync(fullPath)).toBeTruthy();
        expect(data).toMatchSnapshot();
    });
    describe("< 속성 검사 >", () => {
        it("- data : entity.json", () => {
            const prop = changeFakePath(autoTask.entry.data["entity"].getObject());
            expect(JSON.stringify(prop, null, "\t")).toMatchSnapshot();
        });
        it("- helper : bold.js", () => {
            const prop = changeFakePath(autoTask.entry.helper["bold"].getObject());
            expect(JSON.stringify(prop, null, "\t")).toMatchSnapshot();
        });
        it("- part : inc/content.hbs", () => {
            const prop = changeFakePath(autoTask.entry.part["inc/content"].getObject());
            expect(JSON.stringify(prop, null, "\t")).toMatchSnapshot();
        });
        it("- part : inc/foote.hbs", () => {
            const prop = changeFakePath(autoTask.entry.part["inc/footer"].getObject());
            expect(JSON.stringify(prop, null, "\t")).toMatchSnapshot();
        });
        it("- part : inc/header.hbs", () => {
            const prop = changeFakePath(autoTask.entry.part["inc/header"].getObject());
            expect(JSON.stringify(prop, null, "\t")).toMatchSnapshot();
        });
        it("- src : one.html.hbs", () => {
            const prop = changeFakePath(autoTask.entry.src["one.html"].getObject());
            expect(JSON.stringify(prop, null, "\t")).toMatchSnapshot();
        });
    });
});

describe("change File", () => {
    it("[ new >> do_clear(1) >> do_publish() >> 파일 수정 >> do_clear() ]", () => {
        autoTask = AutoTask.create(dirname);    // 생성
        autoTask.isLog = false;
        autoTask.do_clear(1);   // 강제 클리어
        autoTask = AutoTask.create(dirname);    // 생성
        autoTask.isLog = false;
        autoTask.do_publish();
        const fullPath = path.join(dirname, "/src/one.html");
        const data = fs.readFileSync(fullPath,"utf-8");
        if(fs.existsSync(fullPath)) new Error("파일이 없습니다."); // 수정 : 없으면 오류
        // 2.수정
        fs.writeFileSync(fullPath, data + " [EDIT]", "utf8");
        // 3.초기화 
        autoTask.do_clear();
    });
    it("- 수정 파일 여부 : src/one.html ", () => {
        const fullPath = path.join(dirname, "/src/one.html");
        const data = fs.readFileSync(fullPath,"utf-8");
        expect(fs.existsSync(fullPath)).toBeTruthy();
        expect(data).toMatch("[EDIT]");
    });
});

describe("rename template.js", () => {
    it("[ 생성 >> do_clear(1) >> do_publish() ]", () => {
        autoTask = AutoTask.create(dirname, "newTemplate.js");
        autoTask.isLog = false;
        autoTask.do_clear(1);   // 강제 초기화 (별도이 스코프!!)
        autoTask = AutoTask.create(dirname, "newTemplate.js");
        autoTask.isLog = false;
        autoTask.do_publish();
    });
    it("- 생성 여부 : src/one.html", () => {
        const fullPath = dirname + "/src/one.html";
        expect(fs.existsSync(fullPath)).toBeTruthy();
    });
    it("- isRename 속성 여부 : AutoTemplate.isRename = true", () => {
        const isRename = autoTask.entry.isRename;
        expect(isRename).toBeTruthy();
    });
});