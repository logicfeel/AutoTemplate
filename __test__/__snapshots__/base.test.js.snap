// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task :: publish < 속성 검사 > - data : entity.json 1`] = `
"{
	"isPublic": true,
	"content": {
		"Table": "1",
		"Tables": [
			10,
			20
		]
	},
	"dir": "@@",
	"area": "DATA",
	"fullPath": "@@/template/data/entity",
	"areaDir": "template/data",
	"subDir": "",
	"subPath": "entity",
	"localDir": "template/data",
	"localPath": "template/data/entity",
	"name": "entity",
	"fileName": "entity.json",
	"filePath": "@@/template/data/entity.json"
}"
`;

exports[`task :: publish < 속성 검사 > - helper : bold.js 1`] = `
"{
	"isPublic": true,
	"dir": "@@",
	"area": "HELPER",
	"fullPath": "@@/template/helper/bold",
	"areaDir": "template/helper",
	"subDir": "",
	"subPath": "bold",
	"localDir": "template/helper",
	"localPath": "template/helper/bold",
	"name": "bold",
	"fileName": "bold.js",
	"filePath": "@@/template/helper/bold.js"
}"
`;

exports[`task :: publish < 속성 검사 > - part : inc/content.hbs 1`] = `
"{
	"isPublic": true,
	"content": "<content><!--part/inc/content-->\\n    내용임\\n</content>",
	"dir": "@@",
	"area": "PART",
	"fullPath": "@@/template/part/inc/content",
	"areaDir": "template/part",
	"subDir": "inc",
	"subPath": "inc/content",
	"localDir": "template/part/inc",
	"localPath": "template/part/inc/content",
	"name": "content",
	"fileName": "content.hbs",
	"filePath": "@@/template/part/inc/content.hbs",
	"saveName": "content",
	"saveDir": "@@/template/part/inc",
	"savePath": "@@/template/part/inc/content",
	"origin": "inc/content"
}"
`;

exports[`task :: publish < 속성 검사 > - part : inc/foote.hbs 1`] = `
"{
	"isPublic": true,
	"content": "<footer><!--part/inc/footer-->\\n    <p>기본 footer. </p>\\n    $entity.Table={{ entity.Table }}$\\n</footer>\\n",
	"dir": "@@",
	"area": "PART",
	"fullPath": "@@/template/part/inc/footer",
	"areaDir": "template/part",
	"subDir": "inc",
	"subPath": "inc/footer",
	"localDir": "template/part/inc",
	"localPath": "template/part/inc/footer",
	"name": "footer",
	"fileName": "footer.hbs",
	"filePath": "@@/template/part/inc/footer.hbs",
	"saveName": "footer",
	"saveDir": "@@/template/part/inc",
	"savePath": "@@/template/part/inc/footer",
	"origin": "inc/footer"
}"
`;

exports[`task :: publish < 속성 검사 > - part : inc/header.hbs 1`] = `
"{
	"isPublic": true,
	"content": "<header><!--part/inc/header-->\\n  # data 삽입 \\n  $entity.Tables.[0]:{{ entity.Tables.[0] }}$\\n\\n  {{#each params}}\\n      {{this}} sss\\n  {{/each}}\\n\\n  # part 삽입\\n  {{> inc/content }}\\n\\n</header>\\n",
	"dir": "@@",
	"area": "PART",
	"fullPath": "@@/template/part/inc/header",
	"areaDir": "template/part",
	"subDir": "inc",
	"subPath": "inc/header",
	"localDir": "template/part/inc",
	"localPath": "template/part/inc/header",
	"name": "header",
	"fileName": "header.hbs",
	"filePath": "@@/template/part/inc/header.hbs",
	"saveName": "header",
	"saveDir": "@@/template/part/inc",
	"savePath": "@@/template/part/inc/header",
	"origin": "inc/header"
}"
`;

exports[`task :: publish < 속성 검사 > - src : one.html.hbs 1`] = `
"{
	"isPublic": true,
	"content": "<one><!--src/one.html-->\\n\\n    {{#>inc/header }}\\n        @inc/header:none@ 없음\\n    {{/inc/header}}\\n\\n    {{> inc/footer }}\\n\\n    @{{#bold }}BOLD{{/bold }}@\\n</one>",
	"dir": "@@",
	"area": "SRC",
	"fullPath": "@@/src/one.html",
	"areaDir": "src",
	"subDir": "",
	"subPath": "one.html",
	"localDir": "src",
	"localPath": "src/one.html",
	"name": "one.html",
	"fileName": "one.html.hbs",
	"filePath": "@@/src/one.html.hbs",
	"saveName": "one.html",
	"saveDir": "@@/src",
	"savePath": "@@/src/one.html",
	"origin": "one.html"
}"
`;

exports[`task :: publish - 파일 비교 : src/one.html 1`] = `
"<one><!--src/one.html-->

<header><!--part/inc/header-->
  # data 삽입 
  $entity.Tables.[0]:10$


  # part 삽입
  <content><!--part/inc/content-->
      내용임
  </content>
</header>

    <footer><!--part/inc/footer-->
        <p>기본 footer. </p>
        $entity.Table=1$
    </footer>

    @<div class="mybold">BOLD</div>@
</one>"
`;
